### PR TITLE
NIAD-857: process messages

### PR DESCRIPTION
**Pull request by n-r-b**
https://gpitbjss.atlassian.net/browse/NIAD-857

- Updated `InboundQueueService` to listen to messages from Mesh Inbound Queue
- Created `InboundQueueConsumerService` to process the message - translate the message from Edifact to Fhir and publish to an outbound queue
- Created `EdifactToFhirService` which uses a placeholder Fhir transaction mapper `ApprovalTransactionMapper` and returns an empty Fhir message
- Created `OutboundQueueService` to publish message to outbound queue

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Transaction;
+import uk.nhs.digital.nhsconnect.lab.results.outbound.queue.GpOutboundQueueService;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+@Slf4j
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
```
Comment (antonlebedjko) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java:21]: Should we use the new way of doing it? Plus to make it consistent across other classes as well

```suggestion
@RequiredArgsConstructor(onConstructor_ = @Autowired)
```

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java [hunk 2]
```diff
@@ -43,7 +45,8 @@
     protected static final int POLL_INTERVAL_MS = 100;
     protected static final int POLL_DELAY_MS = 10;
     private static final int JMS_RECEIVE_TIMEOUT = 500;
-
+    @Rule
+    public Timeout globalTimeout = Timeout.seconds(2);
```
Comment (andrewwa-kainos) [inline src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java:49]: I think this could be achieved with the `org.junit.jupiter.api.Timeout` annotation on the class here, instead of with the old JUnit 4 rules

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java (hunk 1 above)
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java:21]:
```suggestion
@RequiredArgsConstructor(onConstructor_ = @Autowired)
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/fhir/mapper/TransactionMapperConfig.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.context.annotation.Configuration;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.TransactionType;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Configuration
+public class TransactionMapperConfig {
```
Comment (antonlebedjko) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/fhir/mapper/TransactionMapperConfig.java:13]: Just do double check. We are going to use it somewhere, right? Because for now the IntelliJ says to me that we never use that class, possibly it's just my IntelliJ problem.

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/DataToSend.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    private String operationId;
+    @Getter
+    @Setter
+    private TransactionType transactionType;
+
+    public DataToSend setContent(final Parameters parameters) {
+        this.content = parameters;
+        return this;
+    }
+}
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/DataToSend.java:23]: The annotations can be applied to the class here. Lombok will see that there's already a method called `setContent` and won't try to replace it.

```suggestion
@Getter
@Setter
public class DataToSend {

    private Object content;

    private String operationId;

    private TransactionType transactionType;

    public DataToSend setContent(final Parameters parameters) {
        this.content = parameters;
        return this;
    }
}
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    static TransactionType fromCode(String code) {
+        return Stream.of(
+                Arrays.stream(Inbound.values()),
+                Arrays.stream(Outbound.values()))
+                .flatMap(Function.identity())
+                .map(TransactionType.class::cast)
+                .filter(transactionType -> transactionType.getCode().equals(code))
+                .findFirst()
+                .orElseThrow(IllegalArgumentException::new);
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java:17]: concat will mean you don't need to flatMap with the identity function
```suggestion
        return Stream.concat(Arrays.stream(Inbound.values()), Arrays.stream(Outbound.values()))
                .map(TransactionType.class::cast)
                .filter(transactionType -> transactionType.getCode().equals(code))
                .findFirst()
                .orElseThrow(IllegalArgumentException::new);
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.stream.Stream;
+
+public interface TransactionType {
+
+    static TransactionType fromCode(String code) {
+        return Stream.of(
+                Arrays.stream(Inbound.values()),
+                Arrays.stream(Outbound.values()))
+                .flatMap(Function.identity())
+                .map(TransactionType.class::cast)
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java:14]: `Inbound` and `Outbound` are already `TransationType` extensions, so the cast should be unnecessary

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.nhs.digital.nhsconnect.lab.results.outbound.fhir.FhirParser;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java:11]:
```suggestion
@RequiredArgsConstructor(onConstructor_ = @Autowired)
```

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/InboundQueueServiceTest.java [hunk 8]
```diff
@@ -38,42 +46,107 @@
...
+
+    @Test
+    void receiveInboundMessageWithInvalidWorkflowIdThrowsException() throws Exception {
+        when(message.getStringProperty(JmsHeaders.CONVERSATION_ID)).thenReturn(CONVERSATION_ID);
+        when(message.getBody(String.class)).thenReturn("{\"workflowId\":\"INVALID\"}");
+
+        assertThrows(Exception.class, () -> inboundQueueService.receive(message));
+
+        verify(conversationIdService).applyConversationId(CONVERSATION_ID);
+        verify(message, times(0)).acknowledge();
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/InboundQueueServiceTest.java:92]: We can use `never()` instead of `times(0)` to make this more immediately obvious to the reader.

And elsewhere in this PR.

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java (hunk 7 above)
Comment (antonlebedjko) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java:11]: Should we do it in this way?
```suggestion
@RequiredArgsConstructor(onConstructor_ = @Autowired)
```

**Review by andrewwa-kainos — CHANGES_REQUESTED** [review]

**Review by antonlebedjko — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/fhir/mapper/TransactionMapperConfig.java (hunk 3 above)
Comment (andrewwa-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/fhir/mapper/TransactionMapperConfig.java:13]: Spring looks at this to create the `Map<TransactionType, FhirTransactionMapper>` and injects it into `EdifactToFhirService`

Comment (antonlebedjko) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/fhir/mapper/TransactionMapperConfig.java:13]: @andrewwa-kainos I see. Thanks for explaining!

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java (hunk 7 above)
Comment (antonlebedjko) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java:11]: Resolving, because already has been mentioned by Andrew

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java (hunk 1 above)
Comment (antonlebedjko) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundQueueConsumerService.java:21]: Resolving, because already has been mentioned by Andrew

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java [hunk 9]
```diff
@@ -19,18 +23,15 @@
  * the message from the inbound queue to be the same to the message which has been sent.
```
Comment (antonlebedjko) [inline src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java:23]: Should we modify the comments because now they seem to be not relevant anymore?

**Review by antonlebedjko — DISMISSED** [review]

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java (hunk 2 above)
Comment (n-r-b) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java]: I've added the `@Timeout` annotation to class level but have had to increase the timeout limit to 10 seconds. The previous rule wasn't actually being applied

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/DataToSend.java (hunk 4 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/DataToSend.java]: Good spot!

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java (hunk 5 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/TransactionType.java]: Nice.. even better!

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/InboundQueueServiceTest.java (hunk 8 above)
Comment (n-r-b) [reply src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/queue/InboundQueueServiceTest.java]: Sure 👍

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java (hunk 9 above)
Comment (n-r-b) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java]: I've changed this slightly, let me know if it needs modifying more :-)

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Outbound.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package uk.nhs.digital.nhsconnect.lab.results.model.edifact;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
+public enum Outbound implements TransactionType {
+
+    // TO-DO
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Outbound.java:10]: "TO-DO" as part of this story? If not, do we know which story will cover it and can indicate it here?

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java [hunk 11]
```diff
@@ -1,36 +1,37 @@
...
+import java.io.IOException;
+import java.nio.file.Files;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 /**
  * The test EDIFACT message is sent to the MESH mailbox where the adaptor receives
  * inbound transactions. The test waits for the transaction to be processed and compares
- * the message from the inbound queue to be the same to the message which has been sent.
+ * the message from the GP Outbound queue to be the same to the message which has been sent.
```
Comment (andrewwa-kainos) [inline src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java:23]: Is this accurate for the current state of the system? Ultimately the message on the Outbound GP queue will be a converted message, not the same as that which was originally sent, right?

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Transaction.java [hunk 12]
```diff
@@ -10,6 +10,8 @@
     private final ReferenceTransactionNumber referenceTransactionNumber =
         ReferenceTransactionNumber.fromString(extractSegment(ReferenceTransactionNumber.KEY_QUALIFIER));
 
+    // TO-DO insert segments
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Transaction.java:13]: "TO-DO" as part of this story, or do we know which story will fulfil it and can indicate it here?

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class ObjectSerializer {
+    private final FhirParser fhirParser;
+
+    @SneakyThrows
+    public String serialize(Object object) {
+        if (object instanceof IBaseResource) {
+            return fhirParser.encodeToString((IBaseResource) object);
+        }
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java:19]: Just a note for the future, java 16 is expected to introduce [Pattern Matching for instanceof](https://openjdk.java.net/jeps/394), which would let us do something like
```suggestion
        if (object instanceof IBaseResource resource) {
            return fhirParser.encodeToString(resource);
        }
```
Unlikely to be able to use it professionally though until after the next LTS release, 17 :(

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/utils/JmsHeaders.java [hunk 14]
```diff
@@ -1,5 +1,8 @@
 package uk.nhs.digital.nhsconnect.lab.results.utils;
 
 public final class JmsHeaders {
+
+    public static String OPERATION_ID = "OperationId";
+    public static String TRANSACTION_TYPE = "TransactionType";
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/utils/JmsHeaders.java:6]: Sailesh is making `CONVERSATION_ID` final in #12 (going into #11), and these will need to be final too

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundEdifactTransactionHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.fhir.EdifactToFhirService;
+import uk.nhs.digital.nhsconnect.lab.results.inbound.queue.DataToSend;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Transaction;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class InboundEdifactTransactionHandlerTest {
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundEdifactTransactionHandlerTest.java:17]: Are there any other cases we can test here? A single test always looks sus to me, but could be legitimate.

Comment (n-r-b) [reply src/test/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundEdifactTransactionHandlerTest.java]: I've decided to remove `InboundEdifactTransactionHandler.java` for simplicity - for now, as there wasn't a lot going on in there!

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/utils/JmsHeaders.java (hunk 14 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/utils/JmsHeaders.java]: Done - i must have missed this

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Transaction.java (hunk 12 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Transaction.java]: I've removed the TO-DO for now as after the conversation this morning, this may be renamed or even removed in the upcoming stories

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Outbound.java (hunk 10 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Outbound.java]: I'm unsure about whether this is needed or not, so to simplify things i've converted `TransactionType` to an enum and have removed Outbound/Inbound

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java (hunk 13 above)
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/outbound/queue/ObjectSerializer.java:19]: That looks nice! 👍

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java (hunk 11 above)
Comment (n-r-b) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java]: I've re-worded this slightly, hopefully it makes more sense now? This test will be more useful once we have the mapping in place, currently the outcome is an almost empty json file

**Review by andrewwa-kainos — APPROVED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/InterchangeHeader.java [hunk 16]
```diff
@@ -26,7 +26,7 @@
 @AllArgsConstructor
 public class InterchangeHeader extends Segment {
 
-    public static final String KEY = "UNB";
+    protected static final String KEY = "UNB";
```
Comment (saileshp-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/InterchangeHeader.java:29]: Is there a reason why we've got KEY as `protected` instead of using the `getKey` Getter method?

Comment (andrewwa-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/InterchangeHeader.java:29]: It's a constant class variable - wouldn't normally expose them with a method.

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/InterchangeHeader.java [hunk 17]
```diff
@@ -17,7 +17,7 @@
 /**
  * A specialisation of a segment for the specific use case of an interchange header
  * takes in specific values required to generate an interchange header
- * example: UNB+UNOA:2+TES5+XX11+920113:1317+00000002'
+ * example: UNB+UNOA:2+TES5+XX11+920113:1317+00000002'.
```
[the remark this answers is not in the document]
Comment (n-r-b) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/InterchangeHeader.java]: one of the static analysis tools mentioned ending the sentence with a `.` but it's not enforced so i've removed this

**Review by saileshp-kainos — APPROVED** [review]

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/MessageHeaderTest.java [hunk 18]
```diff
@@ -3,58 +3,94 @@
...
-            .isInstanceOf(EdifactValidationException.class)
-            .hasMessage("UNH: Attribute sequenceNumber is required");
+        assertEquals("UNH: Attribute sequenceNumber must be between 1 and 99999999", exception.getMessage());
     }
 
     @Test
-    public void testValidationStatefulMinMaxSequenceNumber() throws EdifactValidationException {
-        var messageHeader = new MessageHeader();
+    void testValidateStatefulSequenceNumberMoreThanMaxValueThrowsException() {
+        final MessageHeader messageHeader = new MessageHeader(1_00_000_000L);
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/MessageHeaderTest.java:64]:
```suggestion
        final MessageHeader messageHeader = new MessageHeader(100_000_000L);
```

**Review by andrewwa-kainos — APPROVED** [review]

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java (hunk 9 above)
Comment (antonlebedjko) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/inbound/InboundUserAcceptanceTest.java]: Yes, seems good now. Thank you!

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java [hunk 19]
```diff
@@ -136,6 +161,16 @@ protected Message getInboundQueueMessage() {
         return waitFor(() -> jmsTemplate.receive(meshInboundQueueName));
```
Comment (antonlebedjko) [inline src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java:161]: Should we remove this method as we don't use it anymore. What do you think?

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationTestsExtension.java [hunk 20]
```diff
@@ -16,6 +16,7 @@
 import java.util.Objects;
 
 import static org.springframework.jms.support.destination.JmsDestinationAccessor.RECEIVE_TIMEOUT_NO_WAIT;
+import static uk.nhs.digital.nhsconnect.lab.results.IntegrationBaseTest.DLQ_PREFIX;
```
Comment (antonlebedjko) [inline src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationTestsExtension.java:19]: It seems that this import is not used

Comment (antonlebedjko) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationTestsExtension.java:19]:
```suggestion
```

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java (hunk 19 above)
Comment (n-r-b) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java:161]: Oh yes, we should remove this - done

**Review by antonlebedjko — CHANGES_REQUESTED** [review]

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java (hunk 19 above)
Comment (antonlebedjko) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationBaseTest.java]: Perfect. Thanks!

File: src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationTestsExtension.java (hunk 20 above)
Comment (n-r-b) [reply src/intTest/java/uk/nhs/digital/nhsconnect/lab/results/IntegrationTestsExtension.java:19]: Ah, this should be used actually and the declared constant (duplicate) should be removed. I've fixed this now

**Review by antonlebedjko — APPROVED** [review]

**Review by bartek-sarul — APPROVED** [review]

**Review by saileshp-kainos — APPROVED** [review]
